File: sorting/CountSort.java
import java.util.Arrays;

public class CountSort {
    public static void main(String[] args){
        String s = "asdklfjhasdlasdflkjasdfasdfkljasdfkjlkzxcvxzcvqwerwerouiyoiuy";
        System.out.println(s);
        System.out.println(new CountSort().countSortStrings(s));
        int[] nums = {1,2,3,4,2,1,3,4,8,9,7,6,2,3,8,7,9,1,0,2,3,4,7,8,9,0,2,8,3,1,4,6,7,2,9,3,1,0,8,4,7,1,2,3,4,9,8,7};
        System.out.println(Arrays.toString(nums));
        System.out.println(Arrays.toString(new CountSort().countSortNumbers(nums)));

    }
    
    public String countSortStrings(String s){
        StringBuffer result = new StringBuffer();
        int[] count = new int[26];
        for(int i=0; i<s.length(); i++)
            count[s.charAt(i)-'a']++;
        for(int i=0; i<count.length; i++){
            while(count[i]>0){
                char c = (char)('a'+i);
                result.append(c);
                count[i]--;
            }
        }
        return result.toString();
    }

    public int[] countSortNumbers(int[] nums){
        // First we find the Max value
        int max = Integer.MIN_VALUE;
        for(int i=0; i<nums.length; i++)
            if (nums[i]>max)
                max = nums[i];

        // Second we create an array of size that max+1
        int[] arr = new int[max+1];

        // Third we iterate over the nums provided and increment the relavant index
        for(int n: nums)
            arr[n]++;

        // Fourth we iterate over the counting array and create the output
        int[] result = new int[nums.length];
        int count = 0;
        for(int i=0; i< arr.length; i++){
            while(arr[i]>0){
                result[count++] = i;
                arr[i]--;
            }
        }

        return result;
    }
}

File: sorting/CountSort.py
def countSortNumbers(lst):
    # find the max number in the list
    max = lst[0]
    for i in lst:
        if i>max:
            max = i

    # create a counting list and initialize it
    count = [0] * (max+1)
    for i in lst:
        count[i]=count[i]+1;
    
    # iterate over count and create the result
    result = []
    for i in range(0, len(count)):
        while count[i]>0:
            result.append(i)
            count[i]=count[i]-1
    
    return result

def countSortString(str):
    count = [0] * 26
    for ch in str:
        index = ord(ch)-ord('a')
        count[index] = count[index]+1
    
    result = ""
    for i in range(0, len(count)):
        while count[i]>0:
            c = chr(ord('a')+i)
            result+= c
            count[i] = count[i]-1
    return result

test = [1, 2, 3, 4, 2, 1, 3, 4, 8, 9, 7, 6, 2, 3, 8, 7, 9, 1, 0, 2, 3, 
        4, 7, 8, 9, 0, 2, 8, 3, 1, 4, 6, 7, 2, 9, 3, 1, 0, 8, 4, 7, 1, 2, 3, 4, 9, 8, 7]
print(test)
print(countSortNumbers(test))

testString = "asdklfjhasdlasdflkjasdfasdfkljasdfkjlkzxcvxzcvqwerwerouiyoiuy"
print(testString)
print(countSortString(testString))

    
